File: BoardGames/StatsManager.cpp
#include <fstream>
#include <iostream>
#include <vector>
#include "StatsManager.hpp"
#include "Player.hpp"


StatsManager::StatsManager(const std::string& file) : filename(file) {

}


void StatsManager::LoadStats(std::vector <std::shared_ptr<Player>>& currentPlayers) {
	allPlayers.clear();


	std::ifstream fin(filename);
	if (!fin.is_open()) {
		std::ofstream fout("game_stats.txt");
		return;
	}

	// Получаем статистику побед из файла
	std::string name;
	int wins;
	while (fin >> name >> wins) {
		auto old = std::make_shared<Player>(name);
		old->SetStats(wins);
		allPlayers.push_back(old);
	}
	fin.close();

	// Добавление текущих для вывода статистики
	for (auto& current : currentPlayers) {
		bool found = false;
		for (auto& saved : allPlayers) {
			if (saved->GetName() == current->GetName()) {
				current->SetStats(saved->GetStats());
				found = true;
				break;
			}
		}

		if (!found) allPlayers.push_back(current);
	}
}


void StatsManager::SaveStats(const std::vector<std::shared_ptr<Player>>& currentPlayers) {
	// Обновление статистики в списке всех игроков
	for (auto& current : currentPlayers) {
		for (auto& saved : allPlayers) {
			if (saved->GetName() == current->GetName()) {
				saved->SetStats(current->GetStats());
				break;
			}
		}
	}

	std::ofstream fout(filename);
	if (!fout.is_open()) {
		std::cout << "Ошибка открытия файла.\n";
		return;
	}

	// Сохранение статистики в файл
	std::string name;
	int wins;
	for (auto& player : allPlayers) {
		fout << player->GetName() << " " << player->GetStats() << "\n";
	}

	fout.close();
}


void StatsManager::ShowStats() {
	std::cout << "--- Статистика побед ---\n";
	for (auto& player : allPlayers) {
		std::cout << player->GetName() << ": " << player->GetStats() << " побед\n";
	}
}

File: BoardGamesSharp/Core/StatsManager.cs
using System;
using System.Collections.Generic;
using System.Linq;
using System.Text;
using System.Threading.Tasks;

namespace BoardGamesSharp.Core;

internal class StatsManager
{
    private string filename;
    private List<Player> allPlayers = new List<Player>();

    public StatsManager(string file)
    {
        this.filename = file;
    }

    public void LoadStats(List<Player> currentPlayers)
    {
        allPlayers.Clear();

        if (File.Exists(filename))
        {
            using (StreamReader fin = new StreamReader(filename))
            {
                string line;
                while ((line = fin.ReadLine()) != null)
                {
                    string[] lines = line.Split(' ');
                    string name = lines[0];
                    Player oldPlayer = new Player(name);
                    int.TryParse(lines[1], out int wins);
                    oldPlayer.Wins = wins;
                    allPlayers.Add(oldPlayer);
                }
            }
        }


        foreach (Player current in currentPlayers)
        {
            bool found = false;
            foreach (Player saved in allPlayers)
            {
                if (saved.Name == current.Name)
                {
                    current.Wins = saved.Wins;
                    found = true;
                    break;
                }
            }

            if (!found) allPlayers.Add(current);
        }
    }


    public void SaveStats(List<Player> currentPlayers)
    {
        foreach (Player current in currentPlayers)
        {
            foreach (Player saved in allPlayers)
            {
                if (saved.Name == current.Name)
                {
                    saved.Wins = current.Wins;
                    break;
                }
            }
        }

        using (StreamWriter fout = new StreamWriter(filename))
        {
            foreach (Player player in allPlayers)
            {
                fout.WriteLine($"{player.Name} {player.Wins}");
            }
        }
    }


    public void ShowStats()
    {
        Console.WriteLine("--- Статистика побед ---");
        foreach (Player player in allPlayers)
        {
            Console.WriteLine($"{player.Name}: {player.Wins} побед");
        }
    }
}
